### PR TITLE
chore(CRT-281): add function that ensures KubeFedCluster CRD existence in the cluster

### PR DESCRIPTION
**Pull request by MatousJobanek**
We need to create KubeFedCluster CRD programmatically because we cannot install it via CSV - for more information see this issue: https://github.com/operator-framework/operator-lifecycle-manager/issues/1060
The function checks if the CRD already exists, if not then it creates it.

This PR depends on https://github.com/codeready-toolchain/api/pull/43

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -55,6 +56,8 @@ require (
 	sigs.k8s.io/testing_frameworks v0.1.0 // indirect
 )
 
+replace github.com/codeready-toolchain/api => github.com/MatousJobanek/api v0.0.0-20191008135136-1338bbea618f
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply go.mod]: updated in 9cb9675

File: go.sum [hunk 2]
```diff
@@ -2,6 +2,8 @@ cloud.google.com/go v0.34.0 h1:eOI3/cP2VTU6uZLDYAoic+eyzzB9YyGmJ7eIjl8rOPg=
 cloud.google.com/go v0.34.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
 github.com/BurntSushi/toml v0.3.1 h1:WXkYYl6Yr3qBf1K79EBnL4mak0OimBfB0XUf9Vl28OQ=
 github.com/BurntSushi/toml v0.3.1/go.mod h1:xHWCNGjB5oqiDr8zfno3MHue2Ht5sIBksp03qcyfWMU=
+github.com/MatousJobanek/api v0.0.0-20191008135136-1338bbea618f h1:H06CSIL/t13A+JFl7j+sK6LC2dIPgviS+9EeWvmXclo=
+github.com/MatousJobanek/api v0.0.0-20191008135136-1338bbea618f/go.mod h1:/Is7p+crldTqYG0f8LQL7OqmdQRnMJ7u//N82tQdf28=
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply go.sum]: updated in 9cb9675

File: pkg/cluster/kubefedcluster_crd.go [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/types"
+	"os"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+func EnsureKubeFedClusterCrd(log logr.Logger, mgr manager.Manager, stopChan <-chan struct{}) {
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/cluster/kubefedcluster_crd.go]: added as part of b631ea4

File: pkg/cluster/kubefedcluster_crd.go [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+	go func() {
+		if ok := mgr.GetCache().WaitForCacheSync(stopChan); !ok {
+			log.Error(fmt.Errorf("failed to wait for cache to be synced"), "")
+			os.Exit(1)
+		}
+		// create or update all NSTemplateTiers on the cluster at startup
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/cluster/kubefedcluster_crd.go]: removed as part of b631ea4

File: pkg/cluster/kubefedcluster_crd.go [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+	err := client.Get(context.TODO(), types.NamespacedName{Name: "kubefedclusters.core.kubefed.k8s.io"}, crd)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			decoder := serializer.NewCodecFactory(scheme).UniversalDeserializer()
+			kubeFedCrd := &v1beta1.CustomResourceDefinition{}
+			_, _, err := decoder.Decode([]byte(kubeFedClusterCrd), nil, kubeFedCrd)
+			if err != nil {
+				return errs.Wrap(err, "unable to decode the KubeFedCluster CRD")
+			}
+			err = client.Create(context.TODO(), kubeFedCrd)
```
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply pkg/cluster/kubefedcluster_crd.go]: good point. No idea why I chose so complicated approach - probably a lack of oxygen in my brain while implementing it :-)

File: go.sum (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexeykazakov) [reply go.sum]: Still here! :)

File: pkg/cluster/kubefedcluster_crd.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+		return errs.Wrap(err, "unable to decode the KubeFedCluster CRD")
+	}
+	err = client.Create(context.TODO(), kubeFedCrd)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return errs.Wrap(err, "unable to create the KubeFedCluster CRD")
+	}
+	return nil
+}
+
+const kubeFedClusterCrd = `
```
Comment (xcoulon) [inline pkg/cluster/kubefedcluster_crd.go]: we could have externalized this in a .yaml file so we could check its format with the yaml linter.

File: pkg/cluster/kubefedcluster_crd.go [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// EnsureKubeFedClusterCrd creates a KubeFedCluster CRD in the cluster.
+// If the creation returns an error that is of the type "AlreadyExists" then the error is ignored,
+// if the error is of another type then it is returned
+func EnsureKubeFedClusterCrd(scheme *runtime.Scheme, client client.Client) error {
+	decoder := serializer.NewCodecFactory(scheme).UniversalDeserializer()
```
Comment (xcoulon) [inline pkg/cluster/kubefedcluster_crd.go]: we could name this func `EnsureKubeFedClusterCRD` to keep the `CRD` uppercase

**Pull request merged**